### PR TITLE
DotLambda: add parser recovery

### PR DESCRIPTION
**Pull request by DedSec256**
This PR adds parser recovery for the `SynExpr.DotLambda` for cases `_.` and `_`

Special thanks to @auduchinok for the consultation.

### CODE REVIEW

File: src/Compiler/pars.fsy [hunk 1]
```diff
@@ -4891,9 +4891,28 @@ atomicExpr:
...
+        let trivia: SynExprDotLambdaTrivia = { UnderscoreRange = mUnderscore ; DotRange = Some mDot  }
+        SynExpr.DotLambda(expr, unionRanges mUnderscore  expr.Range, trivia), false }
+
+  | UNDERSCORE DOT recover %prec dot_lambda
+      { let mUnderscore = rhs parseState 1
+        let mDot = rhs parseState 2
+        parseState.LexBuffer.CheckLanguageFeatureAndRecover LanguageFeature.AccessorFunctionShorthand (unionRanges mUnderscore mDot)
+        let expr = arbExpr ("dotLambda1", mDot.EndRange)
+        let trivia: SynExprDotLambdaTrivia = { UnderscoreRange = mUnderscore ; DotRange = Some mDot  }
+        SynExpr.DotLambda(expr, unionRanges mUnderscore expr.Range, trivia), false }
```
Comment (auduchinok) [inline src/Compiler/pars.fsy:4903]: It seems `unionRanges` calls may be extracted into something like `mWhole`.

File: tests/service/ParserTests.fs [hunk 2]
```diff
@@ -368,3 +369,65 @@ match () with
     match exprs with
     | SynExpr.Match(_, _, [_; _], _, _) -> ()
     | _ -> failwith "Unexpected tree"
+
+
+[<Test>]
+let ``Dot lambda - Recovery _.`` () =
```
Comment (auduchinok) [inline tests/service/ParserTests.fs:375]: Is there any reason not to use `SyntaxTree` tests and do the manual checks on the tree? 🙂

File: src/Compiler/pars.fsy [hunk 3]
```diff
@@ -4891,9 +4891,28 @@ atomicExpr:
...
-        
+        let trivia: SynExprDotLambdaTrivia = { UnderscoreRange = mUnderscore ; DotRange = Some mDot  }
+        SynExpr.DotLambda(expr, unionRanges mUnderscore  expr.Range, trivia), false }
+
+  | UNDERSCORE DOT recover %prec dot_lambda
+      { let mUnderscore = rhs parseState 1
+        let mDot = rhs parseState 2
+        parseState.LexBuffer.CheckLanguageFeatureAndRecover LanguageFeature.AccessorFunctionShorthand (unionRanges mUnderscore mDot)
+        let expr = arbExpr ("dotLambda1", mDot.EndRange)
+        let trivia: SynExprDotLambdaTrivia = { UnderscoreRange = mUnderscore ; DotRange = Some mDot  }
```
Comment (T-Gro) [inline src/Compiler/pars.fsy:4902]: I would expect a parser error here, what am I missing?

**Review by T-Gro — APPROVED** [review]
Thank you for this change.
I would also prefer to have these tests as syntaxtree tests.

Reason is, if anything changes in the SyntaxTree, the baselines can be easily re-created, whereas specific assertions need to be manually checked.

(which is not a big deal when there is 2 of them, but it is when we have hundreds)

File: src/Compiler/pars.fsy (hunk 3 above)
Comment (auduchinok) [reply src/Compiler/pars.fsy:4902]: `recover` will report an error in face of a wrong token, so this is fine in the most cases. However, a special case is needed for `EOF` and it isn't handled in this rule. Thanks for catching!

**Comment by DedSec256** [comment]
Interesting observations on the commit https://github.com/dotnet/fsharp/pull/16238/commits/27e8d6527e0c6d32f236a4992907981e359c378d:

After adding recovery, in the following invalid code DotLambdas began to be parsed

```fsharp
let a = ( upcast _ ) : obj
```
image



```fsharp
let b = ( _ :> _ ) : obj
```
image



```fsharp
let c = ( _ :> obj)
```
image


Also parens expressions began to be parsed more correctly.

File: src/Compiler/pars.fsy [hunk 4]
```diff
@@ -4891,9 +4891,33 @@ atomicExpr:
...
+        SynExpr.DotLambda(expr, unionRanges mUnderscore  expr.Range, trivia), false }
+
+  | UNDERSCORE DOT recover %prec dot_lambda
+      { let mUnderscore = rhs parseState 1
+        let mDot = rhs parseState 2
+        let mWhole = unionRanges mUnderscore mDot
+        parseState.LexBuffer.CheckLanguageFeatureAndRecover LanguageFeature.AccessorFunctionShorthand mWhole
+
+        let mExpr = mDot.EndRange
+        if not $3 then reportParseErrorAt mExpr (FSComp.SR.parsUnexpectedEndOfFileExpression())
```
Comment (auduchinok) [inline src/Compiler/pars.fsy:4904]: Do we want this error to be at the empty range? Would it be better to cover at least the dot?

File: src/Compiler/pars.fsy [hunk 5]
```diff
@@ -4891,9 +4891,33 @@ atomicExpr:
...
+        let expr = arbExpr ("dotLambda1", mExpr)
+        let trivia: SynExprDotLambdaTrivia = { UnderscoreRange = mUnderscore ; DotRange = Some mDot  }
+        SynExpr.DotLambda(expr, mWhole, trivia), false }
+
+  | UNDERSCORE recover %prec dot_lambda
+      { let mUnderscore = rhs parseState 1
+        parseState.LexBuffer.CheckLanguageFeatureAndRecover LanguageFeature.AccessorFunctionShorthand mUnderscore
+
+        let mExpr = mUnderscore.EndRange
+        if not $2 then reportParseErrorAt mExpr (FSComp.SR.parsUnexpectedEndOfFileExpression())
```
Comment (auduchinok) [inline src/Compiler/pars.fsy:4915]: Likewise, about the empty range.

**Comment by T-Gro** [comment]
@DedSec256 :

Seeing the newly appeared `_` recoveries, I would be inclined to only do a recovery for `UNDERSCORE DOT ..`, but not for a standalone `UNDERSCORE` only.
Chance of breaking something existing (also precedence rules) would decrease dramatically.

File: src/Compiler/pars.fsy [hunk 6]
```diff
@@ -4891,9 +4891,33 @@ atomicExpr:
...
+        parseState.LexBuffer.CheckLanguageFeatureAndRecover LanguageFeature.AccessorFunctionShorthand mWhole
+
+        let mExpr = mDot.EndRange
+        if not $3 then reportParseErrorAt mExpr (FSComp.SR.parsUnexpectedEndOfFileExpression())
+
+        let expr = arbExpr ("dotLambda1", mExpr)
+        let trivia: SynExprDotLambdaTrivia = { UnderscoreRange = mUnderscore ; DotRange = Some mDot  }
+        SynExpr.DotLambda(expr, mWhole, trivia), false }
+
+  | UNDERSCORE recover %prec dot_lambda
```
Comment (T-Gro) [inline src/Compiler/pars.fsy:4910]: I would now be inclined to removing this rule, and only do the recovery if `_.` is typed, incl. the dot.

**Comment by auduchinok** [comment]
> Seeing the newly appeared _ recoveries, I would be inclined to only do a recovery for UNDERSCORE DOT .., but not for a standalone UNDERSCORE only.

Why, though? There're still errors on the incorrect code, so we aren't changing how any correct code is parsed, but surroundings are much better parsed in all the unfinished code examples.

**Comment by T-Gro** [comment]
> > Seeing the newly appeared _ recoveries, I would be inclined to only do a recovery for UNDERSCORE DOT .., but not for a standalone UNDERSCORE only.
> 
> Why, though? There're still errors on the incorrect code, so we aren't changing how any correct code is parsed, but surroundings are much better parsed in all the unfinished code examples.
> 
> The purpose of the parser recovery is to parse as much code in the most correct way as we can, so all the analysis works with the best possible tree. We don't have any ambiguity here, so it's clear that `_` is an unfinished dot lambda, and being able to analyze things like parens or other pars of unfinished code is crucial for the tooling.

I do not see a standalone `_` as a lambda, e.g. in this example `let a = ( upcast _ ) : obj` from @DedSec256 above.

**Comment by auduchinok** [comment]
> I do not see a standalone _ as a lambda, e.g. in this example let a = ( upcast _ ) : obj from @DedSec256 above.

Yes, but it should be something parseable when possible. And here it's unambiguously an unfinished dot lambda, as it's the only kind of an expression that could start with `_`.

It's the same kind of recovery as in the most of [these PRs](https://github.com/dotnet/fsharp/pulls?q=is%3Apr+author%3Aauduchinok+is%3Aclosed+parser+) and it's what makes a much better IDE experience.

It's OK to make assumptions about what an unfinished code may become after the the user types more into the editor. Not doing that makes analysis of surrounding code break. Consider `interface` recovery in https://github.com/dotnet/fsharp/pull/15943. We don't know whether it's going to be an `interface ... end` type representation or whether it's an `interface ... with` member declaration. But it doesn't matter much as long as we can provide improved tooling, and being able to parse it as something is the crucial part here.

**Comment by T-Gro** [comment]
> ... And here it's unambiguously an unfinished dot lambda, as it's the only kind of an expression that could start with `_`. When there're more cases in the language it's possible to update these recovery rules, so it's not a problem...

Couldn't it also be an unfinished identifier starting with an underscore?
What is what I would expect here, at least without the DotLambda feature in place.

**Comment by auduchinok** [comment]
> Couldn't it also be an unfinished identifier starting with an underscore?

It's a good example, thanks! I think would be much better to produce an identifier expression indeed and to report an additional error from inside the rule. It's still going to be parsed as an atomic expression, so everything seems good.

**Comment by auduchinok** [comment]
Another option is to produce a generic error expression node covering `_` range. It's maybe even a better option.

**Comment by T-Gro** [comment]
If it were "just parsed" as an `_` identifier, leaving potential error for typechecking.
Wouldn't it then cover all the needs, including completion working OOB for items prefixed with `_` ?

**Comment by auduchinok** [comment]
> If it were "just parsed" as an _ identifier, leaving potential error for typechecking.
Wouldn't it then cover all the needs, including completion working OOB for items prefixed with _ ?

That's what I'd hope for! 🙂

**Comment by auduchinok** [comment]
> > If it were "just parsed" as an _ identifier, leaving potential error for typechecking.
> > Wouldn't it then cover all the needs, including completion working OOB for items prefixed with _ ?
> 
> That's what I'd hope for! 🙂

The problem I see is it'd be really nice if we the tree could also show that this is an error node somehow in this case. Producing just an identifier moves the responsibility to _various features_ to check it at a later stage, potentially duplicating code. The more I think about it, the more I'd prefer it to be parsed as a dotLambda (or error, in the worst case) if features like code completion keep working, at least for now.

Maybe I'd suggest to keep it as is in this PR if we know that code completion still works for `_` prefix, and do another iteration later, when we're not constrained by a close release date (i.e. we still would like to cherry-pick these changes into our release)?

**Comment by auduchinok** [comment]
Maybe `SynExpr.FromParseError(SynExpr.Named(...))` would work, but it needs testing...

File: src/Compiler/SyntaxTree/SyntaxTrivia.fs [hunk 7]
```diff
@@ -79,7 +79,7 @@ type SynExprLambdaTrivia =
 type SynExprDotLambdaTrivia =
     {
         UnderscoreRange: range
-        DotRange: range
+        DotRange: range option
```
[the remark this answers is not in the document]
Comment (DedSec256) [reply src/Compiler/SyntaxTree/SyntaxTrivia.fs:82]: Nice catch!

**Comment by T-Gro** [comment]
@DedSec256 : Thanks for the changes! Are we ready to merge here?

**Review by auduchinok — APPROVED** [review]

**Review by psfinaki — APPROVED** [review]

**Pull request merged**